common/tokenbucket/tb_test.go

```diff
@@ -167,62 +140,132 @@ func (s *TokenBucketSuite) TestPriorityLowRpsEnforced() {
 		if total >= 3 {
 			break
 		}
-		ts.advance(time.Millisecond * 101)
+		ts.Advance(time.Millisecond * 101)
 	}
-	s.Equal(3, total, "Token bucket failed to enforce limit")
-	s.Equal(3, attempts, "Token bucket gave out tokens too quickly")
+	assert.Equal(t, 3, total, "Token bucket failed to enforce limit")
+	assert.Equal(t, 3, attempts, "Token bucket gave out tokens too quickly")
 }
 
-func (s *TokenBucketSuite) TestPriorityTokenBucket() {
-	ts := &mockTimeSource{currTime: time.Now()}
+func TestPriorityTokenBucket(t *testing.T) {
+	ts := clock.NewMockedTimeSource()
 	tb := NewPriorityTokenBucket(2, 100, ts)
 
 	for i := 0; i < 2; i++ {
 		ok2, _ := tb.GetToken(1, 1)
-		s.False(ok2)
+		assert.False(t, ok2)
 		ok, _ := tb.GetToken(0, 10)
-		s.True(ok)
-		ts.advance(time.Millisecond * 101)
+		assert.True(t, ok)
+		ts.Advance(time.Millisecond * 101)
 	}
 
 	for i := 0; i < 2; i++ {
 		ok, _ := tb.GetToken(0, 9)
-		s.True(ok) // 1 token remaining in 1st bucket, 0 in 2nd
+		assert.True(t, ok) // 1 token remaining in 1st bucket, 0 in 2nd
 		ok2, _ := tb.GetToken(1, 1)
-		s.False(ok2)
-		ts.advance(time.Millisecond * 101)
+		assert.False(t, ok2)
+		ts.Advance(time.Millisecond * 101)
 		ok2, _ = tb.GetToken(1, 2)
-		s.False(ok2)
+		assert.False(t, ok2)
 		ok2, _ = tb.GetToken(1, 1)
-		s.True(ok2)
+		assert.True(t, ok2)
 	}
 }
 
-func (s *TokenBucketSuite) TestFullPriorityTokenBucket() {
-	ts := &mockTimeSource{currTime: time.Now()}
+func TestFullPriorityTokenBucket(t *testing.T) {
+	ts := clock.NewMockedTimeSource()
 	tb := NewFullPriorityTokenBucket(2, 100, ts)
 
 	ok2, _ := tb.GetToken(1, 10)
-	s.True(ok2)
+	assert.True(t, ok2)
 
 	for i := 0; i < 2; i++ {
 		ok2, _ := tb.GetToken(1, 1)
-		s.False(ok2)
+		assert.False(t, ok2)
 		ok, _ := tb.GetToken(0, 10)
-		s.True(ok)
-		ts.advance(time.Millisecond * 101)
+		assert.True(t, ok)
+		ts.Advance(time.Millisecond * 101)
 	}
 
 	ok2, _ = tb.GetToken(1, 1)
-	s.False(ok2)
-	ts.advance(time.Millisecond * 101)
+	assert.False(t, ok2)
+	ts.Advance(time.Millisecond * 101)
 	ok2, _ = tb.GetToken(1, 5)
-	s.True(ok2)
-	ts.advance(time.Millisecond * 101)
+	assert.True(t, ok2)
+	ts.Advance(time.Millisecond * 101)
 	ok2, _ = tb.GetToken(1, 15)
-	s.False(ok2)
+	assert.False(t, ok2)
 	ok2, _ = tb.GetToken(1, 10)
-	s.True(ok2)
+	assert.True(t, ok2)
 	ok, _ := tb.GetToken(0, 10)
-	s.True(ok)
+	assert.True(t, ok)
+}
+
+func TestTokenBucketConsume(t *testing.T) {
+	t.Run("consume", func(t *testing.T) {
+		// make sure we don't deadlock inside the test.
+		go panicOnTimeout(time.Minute)
+
+		ts := clock.NewMockedTimeSource()
+		tb := New(10, ts)
+		ok, _ := tb.TryConsume(1)
+		assert.True(t, ok)
+		consumeFinished := make(chan struct{})
+		go func() {
+			success := tb.Consume(1, 2*refillRate)
+			assert.True(t, success, "Consume returned wrong result")
+			close(consumeFinished)
+		}()
+		ts.BlockUntil(1)
+		select {
+		case <-consumeFinished:
+			assert.Fail(t, "Consume returned before refill")
```

Comment:
Nit: It reqiures quite close reading to follow what you're asserting here, or maybe I'm just being dumb. 

You're consuming on line 210 one entry. A second in the goroutine in 214. Why would a rate-limiter with 10 RPS block in a single time interval? I'm sorry, just not entirely following.

Reply:
ah, ok, the blockuntil(1) waits until the timeout (2 * the refil rate) finishes, so you're checking the timeout occurs ?

Reply:
BlockUntil block until at least 1 waiter will be registered. In this case, time.Sleep inside Consume will create this waiter. After it is stopped, I can Advance time, which should unblock time.Sleep and it will run a single iteration of trying to acquire token. So therefore I'm adding backofInterval first to trigger a single loop. And then advancing time further for rifillRate, which will allow to consume 1 token.

I'll add comments around, but that is how tb logic is done. It is weird and is weird to test.

Reply:
yea, it took me a while to understand what `BlockUntil` did. had to read the source to really figure it out :\

seems useful, just... misleading.